### PR TITLE
transport: move append of header and data down to http2 write loop to save garbage

### LINKED ISSUE #3560: Reduce garbage generation in http send of streaming messages
### Use case(s) - what problem will this feature solve?
Where streaming clients send messages under 16KB, `http2Client.Write()` doubles the memory used for serialisation by copying every message into a fresh allocation.

On this line, `hdr` is a 5-byte slice allocated in `msgHeader()`:
https://github.com/grpc/grpc-go/blob/03def2a1f3d4fd22fb566d561195c31330984c29/internal/transport/http2_client.go#L866

### Proposed Solution
Apologies, I don't know the codebase well.  Maybe `loopyWriter` could hold a single 16KB buffer to do the copy just before each frame write?

Or, if `PreparedMsg` allowed the re-use of `hdr`, it could be pre-allocated at 16KB.

### Additional Context
#2432 suggests the above direction - "would allow users to re-use a PreparedMsg, and may save allocations of internal buffers"

Excerpt from profile which prompted me to open this issue:
```
google.golang.org/grpc/internal/transport.(*http2Client).Write
/go/pkg/mod/google.golang.org/grpc@v1.25.1/internal/transport/http2_client.go

  Total:      5.53GB     5.64GB (flat, cum) 28.47%

    839            .          .           		// Add some data to grpc message header so that we can equally 
    840            .          .           		// distribute bytes across frames. 
    841            .          .           		emptyLen := http2MaxFrameLen - len(hdr) 
    842            .          .           		if emptyLen > len(data) { 
    843            .          .           			emptyLen = len(data) 
    844            .          .           		} 
    845       5.30GB     5.30GB           		hdr = append(hdr, data[:emptyLen]...) 
    846            .          .           		data = data[emptyLen:] 
    847            .          .           		df.h, df.d = hdr, data 
```

### PR DESCRIPTION
**Pull request by bboreham**
Fixes #3560 

This code is quite dense, so I broke the change into two parts: the first should be purely moving checks around, and the second is the meat of the change.

### CODE REVIEW

File: internal/transport/controlbuf.go [hunk 1]
```diff
@@ -857,53 +857,56 @@ func (l *loopyWriter) processData() (bool, error) {
...
+	// Compute how much of the header and data we can send within quota and max frame length
+	hSize := min(maxSize, len(dataItem.h))
+	dSize := min(maxSize-hSize, len(dataItem.d))
+	if hSize != 0 {
+		if dSize == 0 {
+			buf = dataItem.h
+		} else {
+			// We can add some data to grpc message header to distribute bytes more equally across frames.
+			// Copy on the stack to avoid generating garbage
+			var localBuf [http2MaxFrameLen]byte
```
Comment (menghanl) [inline internal/transport/controlbuf.go:882]: Won't `maxSize` be bigger enough?
Or move `size:=hSize+dSize` above this, and `[size]byte`?

And +1 for copying on stack.

Comment (bboreham) [reply internal/transport/controlbuf.go:882]: Thanks for review.
The length of an array in Go must be a constant.

**Comment by dfawley** [comment]
@bboreham since this change is intended to affect performance, can you please run our benchmarks before & after and report the change?

**Comment by bboreham** [comment]
I ran the benchmarks.  This is the kind of thing I was aiming at:

```
streaming-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_8192B-respSize_8192B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps        75258        78230     3.95%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op    100307.58     81383.64   -18.87%
           Allocs/op        70.84        68.36    -2.82%
             ReqT/op 493210828.80 512688128.00     3.95%
            RespT/op 493210828.80 512688128.00     3.95%
            50th-Lat     72.738µs      71.46µs    -1.76%
            90th-Lat    173.911µs    164.265µs    -5.55%
            99th-Lat     860.33µs    857.663µs    -0.31%
             Avg-Lat    132.711µs    127.571µs    -3.87%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev
```

Running the larger set of benchmarks given in `main.go`, I found some results highly unstable from one run to the next.  Here is an illustrative example with two runs that should be the same (code on master at a6ab4473c):

```
$ go run benchmark/benchmain/main.go -benchtime=10s -workloads=unconstrained -maxConcurrentCalls=1 -trace=off -compression=gzip  -reqSizeBytes=1,1 -respSizeBytes=1048576 -networkMode=Local
go1.14.2/grpc1.30.0-dev
unconstrained-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1B-respSize_1048576B-compressor_gzip-channelz_false-preloader_false:
Bytes/op: 101601.28649006813	Allocs/op: 37.41949353960503	
Histogram (empty)
Number of requests:  275634	Request throughput:  220507.2 bit/s
Number of responses: 2059	Response throughput: 1.7272143872e+09 bit/s

go1.14.2/grpc1.30.0-dev
unconstrained-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1B-respSize_1048576B-compressor_gzip-channelz_false-preloader_false:
Bytes/op: 108112.36192885028	Allocs/op: 285.6726672898224	
Histogram (empty)
Number of requests:  180811	Request throughput:  144648.8 bit/s
Number of responses: 1059	Response throughput: 8.883535872e+08 bit/s
```

I spent some time looking for data-recording faults in the benchmarks, but didn't find any.  Is there an explanation somewhere of what the "unconstrained" tests set out to do?

**Comment by dfawley** [comment]
> I spent some time looking for data-recording faults in the benchmarks, but didn't find any. Is there an explanation somewhere of what the "unconstrained" tests set out to do?

Interesting.  When I run this benchmark, the first run is about 10% slower than the following runs, but I don't see anything like these results.  Were you able to reproduce this on the master branch on your machine?

The unconstrained benchmark sends/receives continuously on a single stream (per maxConcurrentCalls setting) for the lifetime of the test.

**Comment by bboreham** [comment]
Yes it was on master.

**Comment by bboreham** [comment]
The odd benchmark results I was seeing are fixed by #3623.

**Comment by bboreham** [comment]
I have to say I'm still not happy with the stability of benchmark results, particularly 'unconstrained'.  I can just about believe it's down to the vagaries of scheduling, but...

With the change at #3623 added to master and to this PR, I get:

```
unary-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps        99984        98350    -1.63%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      9229.22      9200.33    -0.31%
           Allocs/op       193.68       191.81    -1.03%
             ReqT/op     79987.20     78680.00    -1.63%
            RespT/op     79987.20     78680.00    -1.63%
            50th-Lat     77.608µs     78.913µs     1.68%
            90th-Lat    119.761µs    120.961µs     1.00%
            99th-Lat     362.39µs    391.642µs     8.07%
             Avg-Lat     99.781µs    101.425µs     1.65%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev

unary-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1B-respSize_1048576B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps         4645         4766     2.60%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op   5070590.24   5053184.84    -0.34%
           Allocs/op       319.18       317.98    -0.63%
             ReqT/op      3716.00      3812.80     2.58%
            RespT/op 3896508416.00 3998010572.80     2.60%
            50th-Lat   1.716802ms   1.769363ms     3.06%
            90th-Lat   2.975319ms   3.002575ms     0.92%
            99th-Lat  10.577785ms   7.255819ms   -31.41%
             Avg-Lat   2.155416ms   2.097954ms    -2.67%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev

unary-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1048576B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps         4489         4397    -2.05%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op   4996043.10   4985598.65    -0.21%
           Allocs/op       316.57       327.08     3.47%
             ReqT/op 3765646131.20 3688470937.60    -2.05%
            RespT/op      3591.20      3517.60    -2.06%
            50th-Lat   1.796204ms   1.849311ms     2.96%
            90th-Lat   3.230115ms   3.503842ms     8.47%
            99th-Lat  10.983481ms  10.019469ms    -8.78%
             Avg-Lat   2.227279ms   2.273852ms     2.09%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev

unary-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1048576B-respSize_1048576B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps         2613         2638     0.96%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op   9891108.67   9823348.85    -0.69%
           Allocs/op       513.65       492.79    -4.09%
             ReqT/op 2191943270.40 2212914790.40     0.96%
            RespT/op 2191943270.40 2212914790.40     0.96%
            50th-Lat    3.22525ms   3.237756ms     0.39%
            90th-Lat   6.097368ms    5.74367ms    -5.80%
            99th-Lat  11.525757ms  12.391613ms     7.51%
             Avg-Lat   3.826954ms    3.79125ms    -0.93%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev

streaming-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1048576B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps         4098         4427     8.03%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op   5014906.17   4985940.73    -0.58%
           Allocs/op       189.73       187.23    -1.05%
             ReqT/op 3437651558.40 3713636761.60     8.03%
            RespT/op      3278.40      3541.60     8.02%
            50th-Lat   1.794296ms   1.743155ms    -2.85%
            90th-Lat   3.743061ms   3.270307ms   -12.63%
            99th-Lat  15.495269ms  13.715299ms   -11.49%
             Avg-Lat   2.439841ms   2.258439ms    -7.43%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev

streaming-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1048576B-respSize_1048576B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps         2843         2725    -4.15%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op   9290552.28   9266842.71    -0.26%
           Allocs/op       322.06       311.31    -3.42%
             ReqT/op 2384881254.40 2285895680.00    -4.15%
            RespT/op 2384881254.40 2285895680.00    -4.15%
            50th-Lat   2.917067ms   2.898741ms    -0.63%
            90th-Lat   5.382924ms    5.24668ms    -2.53%
            99th-Lat  12.526239ms  17.278014ms    37.93%
             Avg-Lat   3.517489ms   3.669553ms     4.32%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev

streaming-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps       169100       167142    -1.16%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      1590.39      1558.84    -2.01%
           Allocs/op        64.12        62.12    -3.12%
             ReqT/op    135280.00    133713.60    -1.16%
            RespT/op    135280.00    133713.60    -1.16%
            50th-Lat     46.095µs     46.294µs     0.43%
            90th-Lat     74.362µs     74.883µs     0.70%
            99th-Lat    183.235µs    184.681µs     0.79%
             Avg-Lat     58.896µs     59.596µs     1.19%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev

streaming-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1B-respSize_1048576B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps         4692         4826     2.86%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op   5017673.53   5005467.48    -0.24%
           Allocs/op       189.45       188.34    -0.53%
             ReqT/op      3753.60      3860.80     2.85%
            RespT/op 3935934873.60 4048342220.80     2.86%
            50th-Lat   1.745937ms   1.701961ms    -2.52%
            90th-Lat   3.166108ms   3.116448ms    -1.57%
            99th-Lat   8.777453ms   7.942758ms    -9.51%
             Avg-Lat   2.131014ms   2.072575ms    -2.74%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev

unconstrained-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps            0            0      NaN%
             SendOps      2752621      3527752    28.16%
             RecvOps      3329081      2211410   -33.57%
            Bytes/op      1043.74      1021.86    -2.11%
           Allocs/op        26.28        24.81    -7.61%
             ReqT/op   2202096.80   2822201.60    28.16%
            RespT/op   2663264.80   1769128.00   -33.57%
            50th-Lat           0s           0s      NaN%
            90th-Lat           0s           0s      NaN%
            99th-Lat           0s           0s      NaN%
             Avg-Lat           0s           0s      NaN%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev

unconstrained-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1B-respSize_1048576B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps            0            0      NaN%
             SendOps      5222192      4611445   -11.70%
             RecvOps         1430         2054    43.64%
            Bytes/op      3189.03      4435.36    39.07%
           Allocs/op        26.97        25.44    -3.71%
             ReqT/op   4177753.60   3689156.00   -11.70%
            RespT/op 1199570944.00 1723020083.20    43.64%
            50th-Lat           0s           0s      NaN%
            90th-Lat           0s           0s      NaN%
            99th-Lat           0s           0s      NaN%
             Avg-Lat           0s           0s      NaN%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev

unconstrained-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1048576B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps            0            0      NaN%
             SendOps         1772         1036   -41.53%
             RecvOps      4902678      5159503     5.24%
            Bytes/op      3636.81      2386.62   -34.37%
           Allocs/op        24.53        22.58    -8.15%
             ReqT/op 1486461337.60 869059788.80   -41.53%
            RespT/op   3922142.40   4127602.40     5.24%
            50th-Lat           0s           0s      NaN%
            90th-Lat           0s           0s      NaN%
            99th-Lat           0s           0s      NaN%
             Avg-Lat           0s           0s      NaN%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev

unconstrained-networkMode_Local-bufConn_false-keepalive_false-benchTime_10s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1048576B-respSize_1048576B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps            0            0      NaN%
             SendOps         2545         1985   -22.00%
             RecvOps         7531         8515    13.07%
            Bytes/op   7079018.46   6974375.60    -1.48%
           Allocs/op       233.76       228.69    -2.14%
             ReqT/op 2134900736.00 1665138688.00   -22.00%
            RespT/op 6317460684.80 7142899712.00    13.07%
            50th-Lat           0s           0s      NaN%
            90th-Lat           0s           0s      NaN%
            99th-Lat           0s           0s      NaN%
             Avg-Lat           0s           0s      NaN%
           GoVersion     go1.14.2     go1.14.2
         GrpcVersion   1.30.0-dev   1.30.0-dev
```

**Comment by menghanl** [comment]
Thanks for the change, and for running the benchmarks. I think the numbers look good enough for now. We will merge this and monitor the benchmark dashboard for anything abnormal, and will revisit this if it causes any regression. Thanks!

**Pull request merged**